File: activity/20200805/RabbitLee/Leaf-Similar_Trees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from functools import reduce

class Solution:
    def leafSimilar(self, root1: TreeNode, root2: TreeNode) -> bool:
        seq1 = self.get_leaf_value_seq(root1)
        seq2 = self.get_leaf_value_seq(root2)
        if len(seq1) != len(seq2):
            return False
        else:
            return reduce(lambda a, b: a and b, [x == y for x, y in zip(seq1, seq2)], True)
        
    def get_leaf_value_seq(self, root: TreeNode) -> List[int]:
        if root is None:
            return []
        if root.left is None and root.right is None:
            return [root.val]
        else:
            return self.get_leaf_value_seq(root.left) \
                + self.get_leaf_value_seq(root.right)

File: activity/20200805/scau-flyingbird/LeafSimilarTrees.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func leafSimilar(root1 *TreeNode, root2 *TreeNode) bool {
	r1 := make([]int, 0)
	r2 := make([]int, 0)
	searchLeaf(root1, &r1)
	searchLeaf(root2, &r2)
	return checkIsSimilar(r1, r2)
}

func searchLeaf(root *TreeNode, m *[]int) {
	if root == nil {
		return
	} else if root.Left == nil && root.Right == nil {
		*m = append(*m, root.Val)
	} else {
		searchLeaf(root.Left, m)
		searchLeaf(root.Right, m)
	}
}

func checkIsSimilar(r1 []int, r2 []int) bool {
	if len(r1) != len(r2) {
		return false
	}
	for i, v := range r1 {
		if r2[i] != v {
			return false
		}
	}
	return true
}
